okhttp/src/test/java/okhttp3/internal/ws/WebSocketOptionsTest.kt

```diff
@@ -0,0 +1,237 @@
+/*
+ * Copyright (C) 2019 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okhttp3.internal.ws
+
+import java.net.ProtocolException
+import okhttp3.Protocol
+import okhttp3.Request
+import okhttp3.Response
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.Assert.fail
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.junit.runners.Parameterized
+
+@RunWith(Parameterized::class)
+internal class WebSocketOptionsTest(
+  private val extension: String,
+  private val expected: WebSocketOptions?,
+  private val expectedExceptionMessage: String?
+) {
+
+  private companion object {
+
+    val NO_COMPRESSION = WebSocketOptions(
+        compressionEnabled = false,
+        contextTakeover = false
+    )
+
+    val COMPRESSION_NO_TAKEOVER = WebSocketOptions(
+        compressionEnabled = true,
+        contextTakeover = false
+    )
+
+    val COMPRESSION_WITH_TAKEOVER = WebSocketOptions(
+        compressionEnabled = true,
+        contextTakeover = true
+    )
+
+    /**
+     * The only possible extension values in request are
+     *
+     * `permessage-deflate; server_no_context_takeover; client_no_context_takeover`
+     *
+     * and
+     *
+     * `permessage-deflate; server_max_window_bits=15; client_max_window_bits=15`
+     *
+     * See [rfc7692#section-7.1](https://tools.ietf.org/html/rfc7692#section-7.1) for details on
+     * negotiation process.
+     */
+    @JvmStatic
+    @Parameterized.Parameters(name = "{0}")
+    fun parameters(): Collection<Array<out Any?>> = listOf(
```

Comment:
You might be able to make each of these parameters into its own test? I don’t think it needs to be NxM?